sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/StateStoreSuite.scala

```diff
@@ -64,6 +64,63 @@ class StateStoreSuite extends StateStoreSuiteBase[HDFSBackedStateStoreProvider]
     require(!StateStore.isMaintenanceRunning)
   }
 
+  test("retaining only latest configured size of versions in memory") {
```

Comment:
Sorry I didn't catch this earlier. We should ideally have tests that directly validate the specific behaviors we're documenting in the conf:

* '2' will read from cache in the direct failure case
* '1' will read from cache in the happy path but not if there's a failure
* '0' will never populate or read from the cache

Reply:
It is fairly easy to check whether reading from cache or reading from file with https://github.com/apache/spark/pull/21469/commits/c9aada520889b87ace0886805910f0d56d099bd2 in #21469 since it introduces metrics for cache hit and cache miss, but not easy to check in this PR itself.

So I just rely on checking cache to ensure the data is correctly evicted and not available in cache as expected. Hope this is OK.

Btw, I caught a silly bug while adding tests to cover your suggestion. Thanks!